src/darwin/Framework/CHIPTests/MTRDeviceTests.m

```diff
@@ -1648,42 +1649,51 @@ - (void)test020_ReadMultipleAttributes
-                                    }
-                                    XCTAssertNotNil(result[@"data"]);
-                                    XCTAssertNil(result[@"error"]);
-                                    XCTAssertTrue([result[@"data"] isKindOfClass:[NSDictionary class]]);
+                        XCTAssertTrue([values isKindOfClass:[NSArray class]]);
+                        NSArray * resultArray = values;
+                        size_t attributeResultCount = 0;
+                        size_t eventResultCount = 0;
+                        for (NSDictionary * result in resultArray) {
+                            if ([result objectForKey:@"eventPath"]) {
+                                ++eventResultCount;
+                                MTREventPath * path = result[@"eventPath"];
+                                XCTAssertEqualObjects(path.endpoint, @0);
+                                XCTAssertEqualObjects(path.cluster, @40);
+                                XCTAssertEqualObjects(path.event, @0);
+                                XCTAssertNotNil(result[@"data"]);
+                                XCTAssertNil(result[@"error"]);
+                                XCTAssertTrue([result[@"data"] isKindOfClass:[NSDictionary class]]);
+                            } else if ([result objectForKey:@"attributePath"]) {
+                                ++attributeResultCount;
+                                MTRAttributePath * path = result[@"attributePath"];
+                                if ([path.attribute unsignedIntegerValue] < 4) {
+                                    XCTAssertEqualObjects(path.cluster, @29);
+                                    __auto_type endpoint = [path.endpoint unsignedShortValue];
+                                    XCTAssertTrue(endpoint == 0 || endpoint == 1 || endpoint == 2);
+                                } else {
+                                    XCTAssertEqualObjects(path.cluster, @40);
+                                    XCTAssertEqualObjects(path.endpoint, @0);
                                 }
+                                XCTAssertNotNil(result[@"data"]);
+                                XCTAssertNil(result[@"error"]);
+                                XCTAssertTrue([result[@"data"] isKindOfClass:[NSDictionary class]]);
+                            } else {
+                                XCTFail("Unexpected result dictionary %@", result);
                             }
-                            XCTAssertTrue(includeEventPath);
-                            XCTAssertTrue([resultArray count] > 0);
                         }
+                        // Our test application has 3 endpoints.  We have a descriptor on each one,
+                        // so that's 4 results per endpoint, and we only have Basic Information on
+                        // endpoint 0, so that's 4 more results.
+                        XCTAssertEqual(attributeResultCount, 3 * 4 + 4);
```

Comment:
Is this going to start failing if more stuff gets added to the test app?

Reply:
If more endpoints get added, then yes.  At that point we would need to adjust the test.

We could fix that by reading the list of endpoints off the device and using its length instead of hardcoding the 3....

Reply:
Done.